File: src/java/infra/src/main/java/com/orders/constructs/InstrumentedFunction.java
package com.orders.constructs;

import org.jetbrains.annotations.NotNull;
import software.amazon.awscdk.Duration;
import software.amazon.awscdk.services.lambda.*;
import software.amazon.awscdk.services.lambda.Runtime;
import software.amazon.awscdk.services.s3.Bucket;
import software.amazon.awscdk.services.s3.IBucket;
import software.amazon.awscdk.services.s3.assets.Asset;
import software.constructs.Construct;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class InstrumentedFunction extends Construct {
    private final IFunction function;
    private final IAlias alias;

    public InstrumentedFunction(@NotNull Construct scope, @NotNull String id, @NotNull InstrumentedFunctionProps props) {
        super(scope, id);

        Asset fileAsset = Asset.Builder.create(this, "LambdaFunctionJarS3Asset")
                .path("../application/target/com.orders.application-0.0.1-SNAPSHOT-aws.jar").build();

        Map<String, String> lambdaEnvironment = new HashMap<>();
        lambdaEnvironment.put("MAIN_CLASS", "com.orders.application.FunctionConfiguration");
        lambdaEnvironment.put("AWS_LAMBDA_EXEC_WRAPPER", "/opt/datadog_wrapper");
        lambdaEnvironment.put("DD_SITE", "datadoghq.eu");
        lambdaEnvironment.put("DD_SERVICE", props.getSharedProps().getService());
        lambdaEnvironment.put("DD_ENV", props.getSharedProps().getEnv());
        lambdaEnvironment.put("ENV", props.getSharedProps().getEnv());
        lambdaEnvironment.put("DD_VERSION", props.getSharedProps().getVersion());
        lambdaEnvironment.put("DD_API_KEY_SECRET_ARN", props.getSharedProps().getDdApiKeySecret().getSecretArn());
        lambdaEnvironment.put("DD_CAPTURE_LAMBDA_PAYLOAD", "true");
        lambdaEnvironment.put("spring_cloud_function_definition", props.getRoutingExpression());

        // Add custom environment variables to the default set.
        lambdaEnvironment.putAll(props.getEnvironmentVariables());

        List<ILayerVersion> layers = new ArrayList<>(2);
        layers.add(LayerVersion.fromLayerVersionArn(this, "DatadogJavaLayer", "arn:aws:lambda:eu-west-1:464622532012:layer:dd-trace-java:15"));
        layers.add(LayerVersion.fromLayerVersionArn(this, "DatadogLambdaExtension", "arn:aws:lambda:eu-west-1:464622532012:layer:Datadog-Extension:59"));

        IBucket bucket = Bucket.fromBucketName(this, "CDKBucket", fileAsset.getS3BucketName());

        // Create our basic function
        this.function = Function.Builder.create(this, "OrderConfirmedHandler")
                .functionName(id)
                .snapStart(SnapStartConf.ON_PUBLISHED_VERSIONS)
                .runtime(Runtime.JAVA_21)
                .memorySize(2048)
                .handler("org.springframework.cloud.function.adapter.aws.FunctionInvoker::handleRequest")
                .environment(lambdaEnvironment)
                .timeout(Duration.seconds(30))
                .code(Code.fromBucket(bucket, fileAsset.getS3ObjectKey()))
                .layers(layers)
                .build();

        var version = new Version(this, String.format("%sVersion", id), VersionProps.builder()
                .lambda(this.function)
                .build());
        
        this.alias = new Alias(this, String.format("%sAlias", id), AliasProps.builder()
                .aliasName("prod")
                .version(version)
                .build());

        props.getSharedProps().getDdApiKeySecret().grantRead(this.function);
    }

    public IFunction getFunction() {
        return function;
    }

    public IAlias getAlias() {
        return alias;
    }
}

File: src/dotnet/cdk/Constructs/InstrumentedFunction.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Amazon.CDK;
using Amazon.CDK.AWS.Lambda;
using Amazon.CDK.AWS.Lambda.DotNet;
using Amazon.CDK.AWS.Logs;
using Amazon.CDK.AWS.SecretsManager;
using Constructs;

namespace ServerlessGettingStarted.CDK.Constructs;

public record FunctionProps(SharedProps Shared, string FunctionName, string ProjectPath, string Handler, Dictionary<string, string> EnvironmentVariables, ISecret DdApiKeySecret);

public class InstrumentedFunction : Construct
{
    public IFunction Function { get; private set; }
    
    public InstrumentedFunction(Construct scope, string id, FunctionProps props) : base(scope, id)
    {
        if (props.Handler.Length > 128)
        {
            throw new Exception(
                "Function handler cannot be greater than 128 chars. https://docs.aws.amazon.com/lambda/latest/api/API_CreateFunction.html#lambda-CreateFunction-request-Handler");
        }
        var functionName = $"{props.Shared.ServiceName}-{props.FunctionName}-{props.Shared.Env}";

        var defaultEnvironmentVariables = new Dictionary<string, string>()
        {
            { "POWERTOOLS_SERVICE_NAME", props.Shared.ServiceName },
            { "POWERTOOLS_LOG_LEVEL", "DEBUG" },
            { "AWS_LAMBDA_EXEC_WRAPPER", "/opt/datadog_wrapper" }, 
            { "DD_SITE", System.Environment.GetEnvironmentVariable("DD_SITE") },
            { "DD_ENV", props.Shared.Env },
            { "ENV", props.Shared.Env },
            { "DD_VERSION", props.Shared.Version },
            { "DD_SERVICE", props.Shared.ServiceName },
            { "DD_API_KEY_SECRET_ARN", props.DdApiKeySecret.SecretArn },
            { "DD_CAPTURE_LAMBDA_PAYLOAD", "true" },
        };
        
        Function = new DotNetFunction(this, id,
            new DotNetFunctionProps
            {
                ProjectDir = props.ProjectPath,
                Handler = props.Handler,
                MemorySize = 1024,
                Timeout = Duration.Seconds(29),
                Runtime = Runtime.DOTNET_8,
                Environment = defaultEnvironmentVariables.Union(props.EnvironmentVariables).ToDictionary(x => x.Key, x => x.Value),
                Architecture = Architecture.ARM_64,
                FunctionName = functionName,
                LogRetention = RetentionDays.ONE_DAY,
                Layers =
                [
                    LayerVersion.FromLayerVersionArn(this, "DDExtension", "arn:aws:lambda:eu-west-1:464622532012:layer:Datadog-Extension-ARM:64"),
                    LayerVersion.FromLayerVersionArn(this, "DDTrace", "arn:aws:lambda:eu-west-1:464622532012:layer:dd-trace-dotnet-ARM:15"),
                ],
            });

        props.DdApiKeySecret.GrantRead(Function);
    }
}